src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,168 @@
+package duke.parser;
+
+import duke.exception.DukeException;
+import duke.storage.Storage;
+import duke.task.*;
+import duke.ui.Ui;
+import duke.command.*;
+
+import java.io.IOException;
+
+/** Parser to interpret and execute commands. */
+public class Parser {
+
+    /**
+     * Interprets user input and creates the appropriate command.
+     *
+     * @param input    User input during runtime.
+     * @param storage  Storage object to save the task list that is updated in this method.
+     * @param taskList Task list object to store the tasks that are created from user command in this method.
+     * @return An appropriate command.
+     * @throws DukeException If there is an error in parsing user input
+     * @throws IOException   If there is an error in saving task list.
+     */
+    public static Command parseUserInput(String input, Storage storage, TaskList taskList) throws DukeException {
+
+        String[] inputArr = input.split(" ");
+        String instruction = inputArr[0];
+
+        Command command = null;
+
+        switch (instruction) {
+
+        case "done": {
+
+            if (inputArr.length <= 1) {
+                throw new DukeException("☹ OOPS!!! Please specify a task number to mark as done!");
+            }
+
+            int taskNumber = Integer.parseInt(inputArr[1]);
+
+            if (taskNumber > taskList.getSize()) {
+                throw new DukeException("☹ OOPS!!! Please specify a valid task number!");
+            }
+
+            command = new DoneCommand(taskNumber);
+
+            break;
+        }
+
+        case "delete": {
+
+            if (inputArr.length <= 1) {
+                throw new DukeException("☹ OOPS!!! Please specify a task number to be deleted!");
+            }
+
+            int taskNumber = Integer.parseInt(inputArr[1]);
+
+            if (taskNumber > taskList.getSize()) {
+                throw new DukeException("☹ OOPS!!! Please specify a valid task number!");
+            }
+
+            command = new DeleteCommand(taskNumber);
+
+            break;
+        }
+
+        case "todo": {
+
+            if (inputArr.length <= 1) {
+                throw new DukeException("☹ OOPS!!! Please enter the descriptions for your to-do list!");
+            }
+
+            // Pre-processing
```

Comment:
I like the placements of comments in the middle of longer implementations of code in order to help readers keep up with the logic

Reply:
I actually have to disagree with this. As far as possible, your code within a single function should be self-explanatory. Otherwise, usually abstraction is the key :)